pyomo/contrib/mindtpy/outer_approximation.py

```diff
@@ -0,0 +1,445 @@
+            # TODO: The logic here is too complicated, can we simplify it?
+            self.process_objective(config,
+                                   move_objective=config.move_objective,
+                                   use_mcpp=config.use_mcpp,
+                                   update_var_con_list=config.add_regularization is None,
+                                   partition_nonlinear_terms=config.partition_obj_nonlinear_terms,
+                                   obj_handleable_polynomial_degree=self.mip_objective_polynomial_degree,
+                                   constr_handleable_polynomial_degree=self.mip_constraint_polynomial_degree)
+            # The epigraph constraint is very "flat" for branching rules.
+            # If ROA/RLP-NLP is activated and the original objective function is linear, we will use the original objective for the main mip.
+            if MindtPy.objective_list[0].expr.polynomial_degree() in self.mip_objective_polynomial_degree and config.add_regularization is not None:
+                MindtPy.objective_list[0].activate()
+                MindtPy.objective_constr.deactivate()
+                MindtPy.objective.deactivate()
+
+            # Save model initial values.
+            self.initial_var_values = list(
+                v.value for v in MindtPy.variable_list)
+
+            # TODO: if the MindtPy solver is defined once and called several times to solve models. The following two lines are necessary. It seems that the solver class will not be init every time call.
+            # For example, if we remove the following two lines. test_RLPNLP_L1 will fail.
+            self.best_solution_found = None
+            self.best_solution_found_time = None
+            self.initialize_mip_problem()
+
+            # Initialization
+            with time_code(self.timing, 'initialization'):
+                self.MindtPy_initialization(config)
+
+            # Algorithm main loop
+            with time_code(self.timing, 'main loop'):
+                self.MindtPy_iteration_loop(config)
+
+            # Load solution
+            if self.best_solution_found is not None:
+                self.load_solution()
+
+            # Get integral info
+            self.get_integral_info()
+
+            config.logger.info(' {:<25}:   {:>7.4f} '.format(
+                'Primal-dual gap integral', self.primal_dual_gap_integral))
+
+        # Update result
+        self.update_result()
+        if config.single_tree:
+            self.results.solver.num_nodes = self.nlp_iter - \
+                (1 if config.init_strategy == 'rNLP' else 0)
+
+        return self.results
+
+    #
+    # Support 'with' statements.
+    #
+    def __enter__(self):
+        return self
+
+    def __exit__(self, t, v, traceback):
+        pass
+
+
+    # iterate.py
+    def MindtPy_iteration_loop(self, config):
```

Comment:
Looks like the beginning (up to line 218) of this method, at least, is also common with the others.

Reply:
I think each algorithm should have its own `MindtPy_iteration_loop` and `solve` methods.

Reply:
I have moved `MindtPy_iteration_loop` and `solve` functions to the base class.